File: javascript/packages/highlighter/test/syntax-renderer.test.ts
import { describe, it, expect, beforeEach } from "vitest"
import { themes } from "../src/themes.js"

import { Range } from "@herb-tools/core"
import { Herb } from "@herb-tools/node-wasm"
import { SyntaxRenderer } from "../src/syntax-renderer.js"
import { ANSI_REGEX } from "../src/ansi.js"

describe("SyntaxRenderer", () => {
  let renderer: SyntaxRenderer

  beforeEach(async () => {
    renderer = new SyntaxRenderer(themes.onedark, Herb)
    await renderer.initialize()
  })

  describe("initialization", () => {
    it("should initialize successfully", async () => {
      const mockHerb = {
        load: async () => { mockHerb.isLoaded = true },
        lex: () => ({ errors: [], value: [] }),
        isLoaded: false,
      }
      const freshRenderer = new SyntaxRenderer(themes.onedark, mockHerb as any)
      expect(freshRenderer.initialized).toBe(false)
      await freshRenderer.initialize()
      expect(freshRenderer.initialized).toBe(true)
    })

    it("should not reinitialize if already initialized", async () => {
      await renderer.initialize()
      expect(renderer.initialized).toBe(true)

      await renderer.initialize()
      expect(renderer.initialized).toBe(true)
    })
  })

  describe("highlight", () => {
    beforeEach(async () => {
      await renderer.initialize()
    })

    it("should throw error if not initialized", async () => {
      const uninitializedHerb = {
        load: async () => {},
        lex: () => ({ errors: [], value: [] }),
        isLoaded: false,
      }
      const uninitializedRenderer = new SyntaxRenderer(themes.onedark, uninitializedHerb as any)

      expect(() => uninitializedRenderer.highlight("<div>test</div>")).toThrow(
        "SyntaxRenderer must be initialized before use",
      )
    })

    it("should return original content for lex errors", async () => {
      const errorHerb = {
        load: async () => {},
        lex: () => ({ errors: ["error"], value: [] }),
        isLoaded: true,
      }

      const errorRenderer = new SyntaxRenderer(themes.onedark, errorHerb as any)
      await errorRenderer.initialize()

      const content = "<invalid>"
      const result = errorRenderer.highlight(content)
      expect(result).toBe(content)
    })

    it("should highlight simple HTML", () => {
      const content = "<div>hello</div>"
      const result = renderer.highlight(content)

      expect(result).toMatchSnapshot()
    })

    it("should handle empty content", () => {
      const result = renderer.highlight("")
      expect(result).toBe("")
    })

    it("should handle content with no tokens", async () => {
      const noTokenHerb = {
        load: async () => {},
        lex: () => ({ errors: [], value: [] }),
        isLoaded: true,
      }

      const noTokenRenderer = new SyntaxRenderer(
        themes.onedark,
        noTokenHerb as any,
      )
      await noTokenRenderer.initialize()

      const content = "plain text"
      const result = noTokenRenderer.highlight(content)

      expect(result).toBe(content)
    })
  })

  describe("theme support", () => {
    it("should work with different themes", async () => {
      const githubLightRenderer = new SyntaxRenderer(themes["github-light"], Herb)
      await githubLightRenderer.initialize()

      const content = "<div>test</div>"
      const result = githubLightRenderer.highlight(content)

      expect(result).toMatchSnapshot()
    })

    it("should work with simple theme", async () => {
      const simpleRenderer = new SyntaxRenderer(themes.simple, Herb)
      await simpleRenderer.initialize()

      const content = "<div>test</div>"
      const result = simpleRenderer.highlight(content)

      expect(result).toMatchSnapshot()
    })
  })

  describe("color disabled mode", () => {
    it("should return plain text when NO_COLOR is set", async () => {
      const originalNoColor = process.env.NO_COLOR
      process.env.NO_COLOR = "1"

      try {
        const noColorRenderer = new SyntaxRenderer(themes.onedark, Herb)
        await noColorRenderer.initialize()

        const content = "<div>test</div>"
        const result = noColorRenderer.highlight(content)

        expect(result).not.toMatch(ANSI_REGEX)
        expect(result).toMatchSnapshot()
      } finally {
        if (originalNoColor === undefined) {
          delete process.env.NO_COLOR
        } else {
          process.env.NO_COLOR = originalNoColor
        }
      }
    })
  })

  describe("ERB content highlighting", () => {
    it("should highlight Ruby keywords in ERB blocks", async () => {
      const erbHerb = {
        load: async () => {},
        lex: () => ({
          errors: [],
          value: [
            { type: "TOKEN_ERB_START", range: Range.from(0, 2) },
            { type: "TOKEN_ERB_CONTENT", range: Range.from(2, 12) },
            { type: "TOKEN_ERB_END", range: Range.from(12, 14) },
          ],
        }),
        isLoaded: true,
      }

      const erbRenderer = new SyntaxRenderer(themes.onedark, erbHerb as any)
      await erbRenderer.initialize()

      const content = "<% if true %>"
      const result = erbRenderer.highlight(content)

      expect(result).toMatchSnapshot()
    })
  })

  describe("comment state tracking", () => {
    it("should track HTML comment state", async () => {
      const commentHerb = {
        isLoaded: true,
        load: async () => {},
        lex: () => ({
          errors: [],
          value: [
            { type: "TOKEN_HTML_COMMENT_START", range: Range.from(0, 4) },
            { type: "TOKEN_IDENTIFIER", range: Range.from(4, 11) },
            { type: "TOKEN_HTML_COMMENT_END", range: Range.from(11, 14) },
          ],
        }),
      }

      const commentRenderer = new SyntaxRenderer(
        themes.onedark,
        commentHerb as any,
      )
      await commentRenderer.initialize()

      const content = "<!-- comment -->"
      const result = commentRenderer.highlight(content)

      expect(result).toMatchSnapshot()
    })

    it("should preserve ERB highlighting in comments", async () => {
      const erbCommentHerb = {
        isLoaded: true,
        load: async () => {},
        lex: () => ({
          errors: [],
          value: [
            { type: "TOKEN_HTML_COMMENT_START", range: Range.from(0, 4) },
            { type: "TOKEN_ERB_START", range: Range.from(5, 7) },
            { type: "TOKEN_ERB_CONTENT", range: Range.from(7, 12) },
            { type: "TOKEN_ERB_END", range: Range.from(12, 14) },
            { type: "TOKEN_HTML_COMMENT_END", range: Range.from(15, 18) },
          ],
        }),
      }

      const erbCommentRenderer = new SyntaxRenderer(
        themes.onedark,
        erbCommentHerb as any,
      )
      await erbCommentRenderer.initialize()

      const content = "<!-- <% code %> -->"
      const result = erbCommentRenderer.highlight(content)

      expect(result).toMatchSnapshot()
    })
  })
})

File: rust/herb-highlighter/tests/syntax_renderer_test.rs
mod common;

use std::sync::Arc;

use herb_highlighter::ansi::find_ansi_sequences;
use herb_highlighter::herb_backend::Herb;
use herb_highlighter::syntax_renderer::SyntaxRenderer;
use herb_highlighter::themes::{get_theme, Theme};

use common::{no_color, token, with_color, StubBackend};

fn renderer(theme: Theme) -> SyntaxRenderer {
  SyntaxRenderer::with_backend(get_theme(theme).clone(), Arc::new(Herb))
}

#[test]
fn returns_original_content_for_lex_errors() {
  with_color();

  let error_renderer = SyntaxRenderer::with_backend(get_theme(Theme::OneDark).clone(), StubBackend::failing());

  let content = "<invalid>";

  assert_eq!(error_renderer.highlight(content), content);
}

#[test]
fn highlights_simple_html() {
  with_color();

  insta::assert_snapshot!(renderer(Theme::OneDark).highlight("<div>hello</div>"));
}

#[test]
fn handles_empty_content() {
  with_color();

  assert_eq!(renderer(Theme::OneDark).highlight(""), "");
}

#[test]
fn handles_content_with_no_tokens() {
  with_color();

  let no_token_renderer = SyntaxRenderer::with_backend(get_theme(Theme::OneDark).clone(), StubBackend::with_tokens(Vec::new()));

  let content = "plain text";

  assert_eq!(no_token_renderer.highlight(content), content);
}

#[test]
fn works_with_different_themes() {
  with_color();

  insta::assert_snapshot!(renderer(Theme::GithubLight).highlight("<div>test</div>"));
}

#[test]
fn works_with_simple_theme() {
  with_color();

  insta::assert_snapshot!(renderer(Theme::Simple).highlight("<div>test</div>"));
}

#[test]
fn returns_plain_text_when_color_is_off() {
  no_color();

  let result = renderer(Theme::OneDark).highlight("<div>test</div>");

  assert!(find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}

#[test]
fn highlights_ruby_keywords_in_erb_blocks() {
  with_color();

  let erb_renderer = SyntaxRenderer::with_backend(
    get_theme(Theme::OneDark).clone(),
    StubBackend::with_tokens(vec![
      token("TOKEN_ERB_START", 0, 2),
      token("TOKEN_ERB_CONTENT", 2, 12),
      token("TOKEN_ERB_END", 12, 14),
    ]),
  );

  insta::assert_snapshot!(erb_renderer.highlight("<% if true %>"));
}

#[test]
fn tracks_html_comment_state() {
  with_color();

  let comment_renderer = SyntaxRenderer::with_backend(
    get_theme(Theme::OneDark).clone(),
    StubBackend::with_tokens(vec![
      token("TOKEN_HTML_COMMENT_START", 0, 4),
      token("TOKEN_IDENTIFIER", 4, 11),
      token("TOKEN_HTML_COMMENT_END", 11, 14),
    ]),
  );

  insta::assert_snapshot!(comment_renderer.highlight("<!-- comment -->"));
}

#[test]
fn preserves_erb_highlighting_in_comments() {
  with_color();

  let erb_comment_renderer = SyntaxRenderer::with_backend(
    get_theme(Theme::OneDark).clone(),
    StubBackend::with_tokens(vec![
      token("TOKEN_HTML_COMMENT_START", 0, 4),
      token("TOKEN_ERB_START", 5, 7),
      token("TOKEN_ERB_CONTENT", 7, 12),
      token("TOKEN_ERB_END", 12, 14),
      token("TOKEN_HTML_COMMENT_END", 15, 18),
    ]),
  );

  insta::assert_snapshot!(erb_comment_renderer.highlight("<!-- <% code %> -->"));
}
